src/json-rpc-socket.ts

```diff
@@ -89,33 +89,61 @@ export class JsonRpcSocket {
    * Sends a JSON-RPC request through the socket and keeps a listener open to read associated responses as they arrive.
    * Returns a close method to clean up the listener.
    */
-  subscribe(request: JsonRpcRequest, listener: (response: JsonRpcResponse) => void): { close: () => void } {
-    request.id ??= uuidv4();
+  async subscribe(request: JsonRpcRequest, listener: (response: JsonRpcResponse) => void): Promise<{
+    response: JsonRpcResponse;
+    close?: () => Promise<void>;
+   }> {
+
+    if (!request.method.startsWith('rpc.subscribe.')) {
+      throw new Error('subscribe rpc requests must include the `rpc.subscribe` prefix');
+    }
+
+    if (!request.subscribe) {
+      throw new Error('subscribe rpc requests must include subscribe options');
+    }
+
+    const subscriptionId = request.subscribe.id;
     const messageHandler = (event: { data: any }):void => {
       const jsonRpcResponse = JSON.parse(event.data.toString()) as JsonRpcResponse;
-      if (jsonRpcResponse.id === request.id) {
-        // if the incoming response id matches the request id, trigger the listener
-        return listener(jsonRpcResponse);
+      if (jsonRpcResponse.id === subscriptionId) {
+        if (jsonRpcResponse.error !== undefined) {
+          // remove the event listener upon receipt of a JSON RPC Error.
+          this.socket.removeEventListener('message', messageHandler);
+          this.closeSubscription(subscriptionId);
+        }
+        listener(jsonRpcResponse);
```

Comment:
Sanity: no need to wrap in `else` to avoid this being called in the case of the initial occurrence of an error triggered by `if (jsonRpcResponse.error !== undefined)`?

Reply:
I intentionally left it so that the listener can also receive the error. The Socket and subscriptions are just cleaning up themselves to not receive anything further, If I understood what you're asking correctly.